src/GuiBase/Viewer/TrackballCamera.cpp

```diff
@@ -161,6 +154,79 @@ bool Gui::TrackballCamera::handleKeyReleaseEvent( QKeyEvent* e ) {
     return false;
 }
 
+void Gui::TrackballCamera::setCamera( Engine::Camera* camera ) {
+    if ( !camera )
+        return;
+    m_camera = camera;
+    m_trackballCenter = m_camera->getPosition() + 2 * m_camera->getDirection().normalized();
+    m_distFromCenter = 2.0;
+    updatePhiTheta();
+    m_camera->show( false );
+}
+
+void Gui::TrackballCamera::save( std::ostream& out ) const {
+    out << "#Radium_camera_state" << std::endl;
+    out << (int)m_camera->getType() << std::endl;
+    out << m_camera->getFrame().matrix() << std::endl;
+    out << std::endl;
+    out << m_camera->getFOV() << " " << m_camera->getZNear() << " " << m_camera->getZFar() << " "
+        << m_camera->getZoomFactor() << " " << m_camera->getAspect() << std::endl;
+    out << std::endl;
+    out << m_trackballCenter.transpose() << " " << m_distFromCenter << std::endl;
+}
+
+void Gui::TrackballCamera::load( std::istream& in ) {
+    std::string str;
+    int type;
+    Scalar M[16]; // 4x4 view matrix;
+    Scalar fov, znear, zfar, zoom, a, x, y, z, d;
+
+    in >> str;
+    bool result = !in.fail();
+    in >> type;
+    result = !in.fail();
+    for ( uint i = 0; i < 16; ++i )
+    {
+        in >> M[i];
+        result &= !in.fail();
+    }
+    in >> fov >> znear >> zfar >> zoom >> a >> x >> y >> z >> d;
+    result &= !in.fail();
+
+    if ( !result )
+    {
+        LOG( logWARNING ) << "Could not load camera file data";
+        return;
+    }
+    m_distFromCenter = d;
+
+    Core::Matrix4 frame;
+    frame << M[0], M[1], M[2], M[3], M[4], M[5], M[6], M[7], M[8], M[9], M[10], M[11], M[12], M[13],
+        M[14], M[15];
+
+    Core::Transform T( frame );
+    m_camera->setType( Engine::Camera::ProjType( type ) );
+    m_camera->setFrame( T );
+    m_camera->setFOV( fov );
+    m_camera->setZNear( znear );
+    m_camera->setZFar( zfar );
+    m_camera->setZoomFactor( zoom );
+    m_trackballCenter = Core::Vector3( x, y, z );
+
+    updatePhiTheta();
+
+    if ( m_hasLightAttached )
+    {
+        m_light->setPosition( m_camera->getPosition() );
+        m_light->setDirection( m_camera->getDirection() );
+    }
+
+    save( std::cout );
+
+    emit cameraPositionChanged( m_camera->getPosition() );
+    emit cameraTargetChanged( m_trackballCenter );
```

Comment:
By sending two successive signals, we take the risk to render twice in a row the scene.
I don't say it is bad, I'm just asking if it is what we want.

Reply:
Should we replace by a signal for when both have to be fired together?

Reply:
Can be an option yes: `cameraChanged` ?